internal/client/state-db/noncanonical.go

```diff
@@ -0,0 +1,670 @@
+			}
+		}
+		if index == -1 {
+			continue
+		}
+		// Check that it does not have any children
+		if levelIndex != levelCount-1 {
+			for _, h := range nco.parents {
+				if h == hash {
+					log.Printf("DEBUG: Trying to remove block %v with children\n", hash)
+					return nil
+				}
+			}
+		}
+		overlay := level.remove(uint(index))
+		nco.levels.Set(levelIndex, level)
+		commit.Meta.Deleted = append(commit.Meta.Deleted, overlay.journalKey)
+		delete(nco.parents, overlay.hash)
+		discardValues(nco.values, overlay.inserted)
+		break
+	}
+	if nco.levels.Len() > 0 && len(nco.levels.Back().blocks) == 0 {
+		nco.levels.PopBack()
+	}
+	if len(commit.Meta.Deleted) > 0 {
+		return &commit
+	}
+	return nil
+}
+
+// Pin state values in memory
+func (nco *NonCanonicalOverlay[BlockHash, Key]) Pin(hash BlockHash) {
+	refs := nco.pinned[hash]
+	if refs == 0 {
+		log.Println("TRACE: Pinned non-canon block:", hash)
+	}
+	refs += 1
+	nco.pinned[hash] = refs
+}
+
+// Unpin will discard pinned state
+func (nco *NonCanonicalOverlay[BlockHash, Key]) Unpin(hash BlockHash) {
+	var removed bool
+	entry, ok := nco.pinned[hash]
+	if ok {
+		entry -= 1
+		if entry == 0 {
+			delete(nco.pinned, hash)
+			removed = true
+		} else {
+			removed = false
+		}
+	}
+
+	if removed {
+		var parent *BlockHash = &hash
+		for parent != nil {
+			hash := *parent
+			parentHash, ok := nco.parents[hash]
+			if !ok {
+				parent = nil
+			} else {
+				parent = &parentHash
+			}
+
+			entry, ok := nco.pinnedInsertions[hash]
+			if ok {
+				entry.count -= 1
+				if entry.count == 0 {
+					delete(nco.pinnedInsertions, hash)
+					log.Println("TRACE: Discarding unpinned non-canon block:", hash)
+					discardValues(nco.values, entry.keys)
+					delete(nco.parents, hash)
+				}
+			} else {
+				break
+			}
+		}
+	}
+}
+
+type overlayLevel[BlockHash Hash, Key Hash] struct {
+	blocks      []blockOverlay[BlockHash, Key]
+	usedIndices uint64
+}
+
+func (ol *overlayLevel[BlockHash, Key]) push(overlay blockOverlay[BlockHash, Key]) {
+	ol.usedIndices = ol.usedIndices | 1<<overlay.journalIndex
+	ol.blocks = append(ol.blocks, overlay)
+}
+
+func (ol *overlayLevel[BlockHash, Key]) availableIndex() uint64 {
+	return uint64(bits.TrailingZeros64(^ol.usedIndices))
+}
+
+func (ol *overlayLevel[BlockHash, Key]) remove(index uint) blockOverlay[BlockHash, Key] {
+	ol.usedIndices = ol.usedIndices & ^(1 << ol.blocks[index].journalIndex)
+	b := ol.blocks[index]
+	ol.blocks = append(ol.blocks[:index], ol.blocks[index+1:]...)
+	return b
+}
+
+func newOverlayLevel[BlockHash Hash, Key Hash]() overlayLevel[BlockHash, Key] {
+	return overlayLevel[BlockHash, Key]{}
+}
+
+type journalRecord[BlockHash Hash, Key Hash] struct {
+	Hash       BlockHash
+	ParentHash BlockHash
+	Inserted   []HashDBValue[Key]
+	Deleted    []Key
+}
+
+func toJournalKey(block uint64, index uint64) []byte {
+	type blockIndex struct {
+		Block uint64
+		Index uint64
+	}
+	return toMetaKey([]byte("noncanonical_journal"), blockIndex{block, index})
```

Comment:
We can define a constant for `[]byte("noncanonical_journal")` similar to `const last_canonical`

Reply:
How do you define constant `[]byte`? Also, this is not repeated anywhere, so I don't see the need to define a package level variable.

Reply:
I mean var like `var lastCanonical = []byte("last_canonical")` at the top of the file, not a `const` my bad

Reply:
Yea I don't think it's required since it's only used in `toJournalKey`.  The `lastCanonical` variable is used in multiple functions so I thought it made sense to have it as package level variable to avoid duplications.